src/Dev/Deprecation.php

```diff
@@ -163,8 +183,59 @@ protected static function get_called_method_from_trace($backtrace, $level = 1)
         if ($level == 4 && ($backtrace[$newLevel]['class'] ?? '') === InjectionCreator::class) {
             $newLevel = $newLevel + 4;
         }
+        // handle noticeWithNoReplacment()
+        foreach ($backtrace as $trace) {
+            if (($trace['class'] ?? '') === Deprecation::class
+                && ($trace['function'] ?? '') === 'noticeWithNoReplacment'
+            ) {
+                $newLevel = $newLevel + 1;
+                break;
+            }
+        }
         $called = $backtrace[$newLevel] ?? [];
-        return ($called['class'] ?? '') . ($called['type'] ?? '') . ($called['function'] ?? '');
+        return $called;
+    }
+
+    private static function isCalledFromSupportedCode(array $backtrace): bool
+    {
+        $called = Deprecation::get_called_from_trace($backtrace, 1);
+        $file = $called['file'] ?? '';
+        if (!$file) {
+            return false;
+        }
+        return Deprecation::fileIsInSupportedModule($file);
+    }
+
+    /**
+     * Check whether a file (path to file) is in a supported module
+     */
+    public static function fileIsInSupportedModule(string $file): bool
+    {
+        // Cache the supported modules list
+        if (count(Deprecation::$supportedModuleDirectories) === 0) {
+            // Manually load the supported modules list rather than use MetaData::getAllRepositoryMetaData()
+            // because we do not want to make a network request which could slow down a website
+            // While there is a small risk of the list being out of date, there is minimal downside to this
+            $path = Path::join(BASE_PATH, 'vendor/silverstripe/supported-modules/repositories.json');
+            if (!file_exists($path)) {
+                throw new RuntimeException('Could not find supported modules list');
+            }
+            $json = json_decode(file_get_contents($path), true);
+            if (is_null($json)) {
+                throw new RuntimeException('Could not parse supported modules list');
+            }
+            $dirs = array_map(fn($module) => "/vendor/{$module['packagist']}/", $json['supportedModules']);
+            // This is a special case for silverstripe-framework when running in CI
+            // Needed because module is run in the root folder rather than in the vendor folder
+            $dirs[] = '/silverstripe-framework/';
```

Comment:
Can we add a check for `SapphireTest::is_running_test()` or an equivalent please?

Reply:
Cannot because `SapphireTest::is_running_test()` is a protected method

Reply:
"or an equivalent"

Reply:
I don't know what the equivalent is. I don't think it exists. The closest I can think of is [MetaData::$isRunningTests](https://github.com/silverstripe/supported-modules/blob/main/src/MetaData.php#L9) ... but that's set by tests/bootstrap.php in silverstripe/supported-modules 

> I would like to avoid having code that exists to account for CI running in a production site.

Is that what exactly what `SapphireTest::is_running_test()` would be?

Reply:
> Isn't that what exactly what SapphireTest::is_running_test() would be?

No - that would be used to check if CI is running or not, so we can conditionally include code based on that condition. That way, adding `silverstripe-framework` would only be done during CI/tests and _not_ in production code.

Reply:
> I don't know what the equivalent is.

It's probably a new public method to replace the protected one.

Reply:
The existing is_running_tests() method has the following in the docblock which makes me inclined to not make this public

``` 
 * Check if test bootstrapping has been performed. Must not be relied on
 * outside of unit tests.
```

I'm not sure what the reasoning for for "Must not be relied on outside of unit tests" is, though presumably someone put some thought in to it.

Deprecation already has check for `!Director::isDev()` meaning that Deprecations will only ever function in a dev environment, so while this line will run on production code, it won't run in a production environment. If you're worried that someone has a "/silverstripe-framework/" path in their project, I could expand this out to "/silverstripe-framework/silverstripe-framework/" which is the path used in github action CI, though there is more chance that would break in the future.

Reply:
Oh weird. Good spotting, didn't see that in the phpdoc.

I still think this is a bad smell.... but it seems like the appetite to avoid it is low, and the fact that it can only run in dev mode drastically lowers its importance so I'm (relunctantly) okay with merging this as is.